Add TLS support to vtadmin GRPC

corbantek opened the pull request:
<!--
  Thank you for your contribution to the Vitess project.
  How to contribute: https://vitess.io/docs/contributing/
  Please first make sure there is an open Issue to discuss the feature/fix suggested in this PR.
  If this is a new feature, please mark the Issue as "RFC".
 -->

<!-- if this PR is Work in Progress please create it as a Draft Pull Request -->

## Description
### Why
- At HubSpot we use TLS GRPC for all our communications
- We are in the process of upgrading to Vitess 23+ and were unable to bringup `vtadmin` without TLS
- This change adds this functionallity

### What
- Reused same utilities from vtctlclient
- Registers vtadmin to use `grpcclientcommon`
- Updates vtadmin proxy dial options to use grpc secure dial options when possible
<!-- A few sentences describing the overall goals of the pull request's commits. -->
<!-- If this is a bug fix and you think the fix should be backported, please write so. -->

## Related Issue(s)
[Feature Request: expose TLS related flags for vtadmin config](https://github.com/vitessio/vitess/issues/14434#top)
<!-- List related issues and pull requests. If this PR fixes an issue, please add it using Fixes #????  -->

## Checklist

-   [X] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [X] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [X] Tests were added or are not required
-   [X] Did the new or modified tests pass consistently locally and on CI?
-   [X] Documentation was added or is not required

## Deployment Notes

<!-- Notes regarding deployment of the contained body of work. These should note any db migrations, etc. -->

corbantek commented:
> @corbantek Can you also add a test for this?

Do you have any specifics on what you'd like to be tested?  I honestly didn't change anything other than what's already being used by Vitess in vtctldclient and others...

Let me poke around in the meantime, but any advice it appreciated.

corbantek commented:
> > @corbantek Can you also add a test for this?
> 
> @dbussink Do you have any specifics on what you'd like to be tested? I honestly didn't change anything other than what's already being used by Vitess in vtctldclient and others...
> 
> Let me poke around in the meantime, but any advice it appreciated.

Nevermind, I tried sticking an agent on this and had it create/add tests for everything I changed (even added some backwards compatible tests for non-TLS)

dbussink commented:
> Nevermind, I tried sticking an agent on this and had it create/add tests for everything I changed (even added some backwards compatible tests for non-TLS)

This generally means you need to clean things up there. It has the tendency to add way too much, violate formatting etc. So it can be a useful starting point, but let's make sure to reduce it then to a minimal useful test. I now see a big wall of text that does way more than it needs to. 

Do any of the other binaries have tests to verify the minimal necessary thing here?

corbantek commented:
> > Nevermind, I tried sticking an agent on this and had it create/add tests for everything I changed (even added some backwards compatible tests for non-TLS)
> 
> This generally means you need to clean things up there. It has the tendency to add way too much, violate formatting etc. So it can be a useful starting point, but let's make sure to reduce it then to a minimal useful test. I now see a big wall of text that does way more than it needs to.
> 
> Do any of the other binaries have tests to verify the minimal necessary thing here?

I actually reviewed what the agent wrote and removed any non-needed stuff.  I actually think all of these tests are viable. 

In short:
- Testing that the args are on the cli command is useful as I've fixed this for HubSpot in the past during vitess upgrades
- I tested both paths (TLS and Non-TLS)

I did leave the `go/vt/vtadmin/vtctldclient/grpc_integration_test.go` as its doing something a little different, but I'd be comfortable remove that one if you'd like.

corbantek commented:
> > Nevermind, I tried sticking an agent on this and had it create/add tests for everything I changed (even added some backwards compatible tests for non-TLS)
> 
> This generally means you need to clean things up there. It has the tendency to add way too much, violate formatting etc. So it can be a useful starting point, but let's make sure to reduce it then to a minimal useful test. I now see a big wall of text that does way more than it needs to.
> 
> Do any of the other binaries have tests to verify the minimal necessary thing here?

Actually re-reviewing some of these test and tracing them, they are crap.  Let me look at this again tomorrow and see if I can figure out a way to actually test.  

Some of these tests looked like they were building up mocks but when I actually traced they were just testing themselves...

I find testing my change to proxy.go the most difficult...

corbantek commented on go/vt/vtadmin/vtctldclient/proxy_test.go:
```diff
@@ -228,3 +232,79 @@ func TestRedial(t *testing.T) {
+
+	// Test that grpcclient.SecureDialOption fails with these files
+	_, err = grpcclient.SecureDialOption(tmpCert.Name(), tmpKey.Name(), "", "", "")
+	assert.Error(t, err, "SecureDialOption should fail with invalid cert/key files")
```
If SecureDialOption somehow changed this would catch that here so we don't have a bogus test below that always passes...

I can remove if you feel its not needed

dbussink replied:
@corbantek Testing for `SecureDialOption` needs to be in the package where that function lives really. Not outside of it here. I'm not really sure how this test is also then related to the original issue fixed?

corbantek replied:
I only added changed these lines in the original PR:
```golang
    tlsOpt, err := grpcclientcommon.SecureDialOption()
	if err != nil {
		return err
	}
```

You asked me to write unit tests for this PRs changes...

The current unit tests for proxy.go already covered this entire file EXCEPT 
```golang
	if err != nil {
		return err
	}
```

This test builds up a scenario where the arguments to SecureDialOption are invalid causing it throw an error and reach `return err` (the only line not covered by tests)

corbantek commented on go/vt/vtadmin/vtctldclient/proxy_test.go:
```diff
@@ -228,3 +232,79 @@ func TestRedial(t *testing.T) {
+	// The error should be from certificate parsing
+	assert.Contains(t, err.Error(), "tls")
+
+	t.Log("Successfully executed the error path on line 116 in proxy.go")
+	t.Log("Error from line 116:", err.Error())
+}
```
I'll remove this logging

corbantek commented:
@dbussink Bumping this for review.  Anything else you'd like me to address?
